src/target/puya.c

```diff
@@ -0,0 +1,171 @@
+
+/* RAM */
+#define RAM_START 0x20000000U
+
+/* RCC */
+#define RCC_BASE  0x40021000U
+#define RCC_ICSCR (RCC_BASE + 0x04U)
+
+/* DBG */
+#define DBG_BASE   0x40015800U
+#define DBG_IDCODE (DBG_BASE + 0x00U)
+
+/*
+ * Flash functions
+ */
+static bool puya_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len);
+static bool puya_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len);
+static bool puya_flash_prepare(target_flash_s *flash);
+static bool puya_flash_done(target_flash_s *flash);
+
+bool puya_probe(target_s *target)
+{
+	uint32_t ram_size = 0U;
+	size_t flash_size = 0U;
+
+	const uint32_t dbg_idcode = target_mem32_read32(target, DBG_IDCODE);
+	if ((dbg_idcode & 0xfff) == 0) {
+		const uint32_t flash_ram_sz = target_mem32_read32(target, FLASH_RAM_SZ);
+		flash_size = ((flash_ram_sz & 7) + 1) * 0x2000;
+		ram_size = (((flash_ram_sz >> 4) & 3) + 1) * 0x800;
+		// TODO: which part families does this actually correspond to?
+		// Tested with a PY32F002AW15U which returns 0x60001000 in IDCODE
+		target->driver = "PY32Fxxx";
+	} else {
+		DEBUG_TARGET("Unknown PY32 device %08" PRIx32 "\n", dbg_idcode);
+		return false;
+	}
+
+	target_add_ram32(target, RAM_START, ram_size);
+	target_flash_s *flash = calloc(1, sizeof(*flash));
+	if (!flash) { /* calloc failed: heap exhaustion */
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return false;
+	}
+	flash->start = FLASH_START;
+	flash->length = flash_size;
+	flash->blocksize = FLASH_PAGE_SIZE;
+	flash->writesize = FLASH_PAGE_SIZE;
+	flash->erase = puya_flash_erase;
+	flash->write = puya_flash_write;
+	flash->prepare = puya_flash_prepare;
+	flash->done = puya_flash_done;
+	flash->erased = 0xffU;
+	target_add_flash(target, flash);
+
+	return false;
+}
+
+static bool puya_flash_prepare(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY1);
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY2);
+
+	int hsi_fs = (target_mem32_read32(flash->t, RCC_ICSCR) >> 13) & 7;
+	if (hsi_fs > 4)
+		hsi_fs = 0;
+	DEBUG_INFO("HSI frequencey selection is %d\n", hsi_fs);
+
+	const uint32_t eppara0 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 0);
+	const uint32_t eppara1 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 4);
+	const uint32_t eppara2 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 8);
+	const uint32_t eppara3 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 12);
+	const uint32_t eppara4 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 16);
+	DEBUG_INFO("PY32 flash timing cal 0: %08" PRIx32 "\n", eppara0);
+	DEBUG_INFO("PY32 flash timing cal 1: %08" PRIx32 "\n", eppara1);
+	DEBUG_INFO("PY32 flash timing cal 2: %08" PRIx32 "\n", eppara2);
+	DEBUG_INFO("PY32 flash timing cal 3: %08" PRIx32 "\n", eppara3);
+	DEBUG_INFO("PY32 flash timing cal 4: %08" PRIx32 "\n", eppara4);
```

Comment:
What use are these to a typical user of this? Why the INFO level? These should, if kept, go in the TARGET debugging log level. Same for the one above about the HSI.

Similarly, please explain what each of these 5 registers mean/do with your choice of variable naming.

Reply:
Changed to `DEBUG_TARGET`. I have been defaulting to `DEBUG_INFO` because I could never actually manage to make `DEBUG_TARGET` work (despite `-Ddebug_output=true`)

As for the naming, this is the exact naming used in the reference manual. I'm not sure how they could be explained any better? These parameters are just copied from the calibration sector to hardware registers, and other than controlling "timing" their use isn't explained in the manual.

Reply:
`DEBUG_TARGET` is a log level that is intentionally never included in the firmware, but which is included in BMDA - you can access the various log levels with the `-v` option which then allows you to get GDB and higher log levels (the firmware only ever includes ERROR, WARN and INFO when built with debug output enabled, the main reason for this is to manage the binary size which is already very large when we do this).

Reply:
`-Ddebug_output=true`? You're doing on-probe diag logging? `DEBUG_TARGET()` is not currently compiled into in-tree firmware builds, and `DEBUG_INFO()` is. You could mangle `src/include/general.h` to override this, but I'm thinking of escalating this to include said level. It's only 3116 bytes more on latest `main` for `blackpill-f411ce` with every driver enabled: 162808-159692 `build-f411ce/blackmagic_blackpill_f411ce_firmware.bin`, and for f103 platforms even less: 120452-118256=2196 `build-swlink/blackmagic_swlink_firmware.bin` (configured as `meson setup --cross-file cross-file/swlink.ini -Drtt_support=true -Ddebug_output=true -Dtargets=cortexm,hc32,lpc,nxp,rp,stm build-swlink/`)